syntax = "proto3";
package code_executor_grpc;

message ChatMessageText {
    string text = 1;
}

message ChatMessageImageUrl {
    string url = 1;
    // ignore detail for now
}

message ChatMessageImage {
    string media_type = 1;
    string data = 2;
}

message ChatMessageContentPart {
    oneof value {
        ChatMessageText text = 1;
        ChatMessageImageUrl image_url = 2;
        ChatMessageImage image = 3;
    }
}

message ContentPartList {
    repeated ChatMessageContentPart parts = 1;
}

message ChatMessageContent {
    oneof value {
        string text = 1;
        ContentPartList content_part_list = 2;
    }
}

message ChatMessageList {
    message ChatMessage {
        string role = 1;
        ChatMessageContent content = 2;
    }

    repeated ChatMessage messages = 1;
}

message StringList {
    repeated string values = 1;
}

message Arg {
    oneof value {
        string string_value = 1;
        ChatMessageList messages_value = 2;
        StringList string_list_value = 3;
        double float_value = 4;
        bool bool_value = 5;
    }
}

enum HandleType {
    ANY = 0;
    STRING = 1;
    STRING_LIST = 2;
    CHAT_MESSAGE_LIST = 3;
    FLOAT = 4;
}

message ExecuteCodeRequest {
    string code = 1;
    string fn_name = 2;
    map<string, Arg> args = 3;
    HandleType return_type = 4;
}

message ExecuteCodeResponse {
    message ErrorMessage {
        string message = 1;
    }
    oneof response {
        Arg result = 1;
        ErrorMessage error = 2;
    }
}

service CodeExecutor {
    rpc Execute(ExecuteCodeRequest) returns (ExecuteCodeResponse);
}
